server.go

```diff
@@ -1,28 +1,168 @@
+	// If context deadline is specified, it will exit earlier
+	// or call ServerDNSCrypt.Shutdown.
+	Shutdown(ctx context.Context) error
+}
+
+// Server is a simple DNSCrypt server implementation
 type Server struct {
-	// ProviderName - DNSCrypt provider name
+	// ProviderName is a DNSCrypt provider name
 	ProviderName string
 
-	// ResolverCert - contains resolver certificate.
+	// ResolverCert contains resolver certificate.
 	ResolverCert *Cert
 
 	// Handler to invoke. If nil, uses DefaultHandler.
 	Handler Handler
+
+	// Shutdown handling
+	// --
+	lock         sync.RWMutex // protects access to all the fields below
+	started      bool
+	wg           sync.WaitGroup            // active workers (servers)
+	tcpListeners map[net.Listener]struct{} // track active TCP listeners
+	udpListeners map[*net.UDPConn]struct{} // track active UDP listeners
+	tcpConns     map[net.Conn]struct{}     // track active connections
 }
 
-// serveDNS - serves DNS response
-func (s *Server) serveDNS(rw ResponseWriter, r *dns.Msg) {
-	if r == nil || len(r.Question) != 1 || r.Response {
-		log.Tracef("Invalid query: %v", r)
+// type check
+var _ ServerDNSCrypt = &Server{}
+
+// prepareShutdown - prepares the server to shutdown:
+// unblocks reads from all connections related to this server
+// marks the server as stopped
+// if the server is not started, returns ErrServerNotStarted
+func (s *Server) prepareShutdown() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if !s.started {
+		log.Info("Server is not started")
+		return ErrServerNotStarted
+	}
+
+	s.started = false
+
+	// Unblock reads for all active tcpConns
+	for conn := range s.tcpConns {
+		_ = conn.SetReadDeadline(longTimeAgo)
+	}
+
+	// These listeners were passed to us from the outside so we cannot close
+	// them here - this is up to the calling code to do that. Instead of that,
+	// we call Set(Read)Deadline to unblock goroutines that are currently
+	// blocked on reading from those listeners.
+
+	// Unblock reads for all active TCP listeners
+	for l := range s.tcpListeners {
+		switch v := l.(type) {
+		case *net.TCPListener:
+			_ = v.SetDeadline(longTimeAgo)
+		}
+	}
+
+	// Unblock reads for all active UDP listeners
+	for l := range s.udpListeners {
+		_ = l.SetReadDeadline(longTimeAgo)
+	}
+
+	return nil
+}
+
+// Shutdown tries to gracefully shutdown the server. It waits until all
+// connections are processed and only after that it leaves the method.
+// If context deadline is specified, it will exit earlier.
+func (s *Server) Shutdown(ctx context.Context) error {
+	log.Info("Shutting down the DNSCrypt server")
+
+	err := s.prepareShutdown()
+	if err != nil {
+		return err
+	}
+
+	// Using this channel to wait until all goroutines finish their work
+	closed := make(chan struct{})
+	go func() {
+		s.wg.Wait()
+		log.Info("Serve goroutines finished their work")
+		close(closed)
+	}()
+
+	// Wait for either all goroutines finish their work
+	// Or for the context deadline
+	var ctxErr error
```

Comment:
Do we really need a new variable here?  Why not just use `err` from above?  It is already proven to be nil.